### PR TITLE
Modify API to allow boxed closures for JsFunctions.

### PR DESCRIPTION
**Pull request by abrodersen**
Fix for issue #264. Feedback is welcome.

### CODE REVIEW

**Comment by nixpulvis** [comment]
Is there a use case for unboxed JsFunctions, other than the tiny memory overhead.

File: src/js/mod.rs [hunk 1]
```diff
@@ -48,26 +49,37 @@ pub(crate) mod internal {
     }
 
     #[repr(C)]
-    pub struct FunctionKernel<T: Value>(pub fn(Call) -> JsResult<T>);
+    pub struct FunctionKernel<T: Value>(pub Box<FnMut(Call) -> JsResult<T>>);
```
Comment (RReverser) [inline src/js/mod.rs]: Might be worth having type parameter `F: FnMut(Call) -> JsResult<T>` to avoid dynamic dispatch.

**Comment by dherman** [comment]
Thanks for this contribution! What I'd like to do over time is have new and/or changed APIs go through RFCs, since they affect the overall design. I don't want to throw process at you since you already did this work but it's definitely worth a good discussion.

It's up to you if you want to write an RFC or not but would you mind at least fleshing out some of the motivation and rationale (basically the kinds of stuff that goes in an RFC)?

**Comment by abrodersen** [comment]
I have created an RFC for this pull request: neon-bindings/rfcs#12

**Comment by rebo** [comment]
Hi, I've tried this pull request, when I call twice a jsFunction defined in this way the runtime crashes with a malloc pointer being freed was not allocated error.

code:

            let foo = String::from("foo");
        let bar_func = JsFunction::new(scope, Box::new(move |call| {
        let msg = foo.clone() + "bar";
        
        Ok(JsString::new(call.scope, &msg).unwrap())
        }))?;
        
        
        let args: Vec<Handle<JsValue>> = vec![];
        bar_func.call(scope,JsUndefined::new(),args.clone());
        bar_func.call(scope,JsUndefined::new(),args.clone());  // crashes on second call, first is fine though

**Comment by abrodersen** [comment]
@rebo Thank you for the feedback! I'll look into it today.

**Comment by abrodersen** [comment]
@rebo Can you try out the updated code? I think it should fix the issue you identified.

**Comment by abrodersen** [comment]
@dherman I've updated this PR with feedback from you and others. Would you consider merging it?

**Comment by rebo** [comment]
Hi @abrodersen , seems to work fine now thanks for fixing.  I have it working with your persistent handles and a neon async task.

**Comment by kjvalencik** [comment]
Closing this until we have a solid RFC and have resolved https://github.com/neon-bindings/rfcs/pull/12#issuecomment-535526184.

**Pull request closed without merging**